sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py

```diff
@@ -0,0 +1,171 @@
+            consumers.append(consumer)
+
+            partition_processor = self._partition_processor_callable(
+                eventhub_name=self._eventhub_name,
+                consumer_group_name=self._consumer_group_name,
+                partition_id=partition_id,
+                checkpoint_manager=CheckpointManager(partition_id, self._eventhub_name, self._consumer_group_name,
+                                                     self._instance_id, self._partition_manager)
+            )
+            loop = asyncio.get_running_loop()
+            task = loop.create_task(
+                _receive(consumer, partition_processor, self._receive_timeout, self._cancellation_token))
+            self._tasks.append(task)
+
+        await asyncio.gather(*self._tasks)
+        await self._partition_manager.close()
+        log.info("EventProcessor %r partition manager is closed", self._instance_id)
+        log.info("EventProcessor %r partition has stopped", self._instance_id)
+
+
+async def _receive(partition_consumer, partition_processor, receive_timeout, cancellation_token):
+    async with partition_consumer:
+        while not cancellation_token.is_cancelled:
+            try:
+                events = await partition_consumer.receive(timeout=receive_timeout)
+            except EventHubError as eh_err:
+                if eh_err.error == "link:stolen":
+                    reason = CloseReason.LEASE_LOST
+                else:
+                    reason = CloseReason.EVENTHUB_EXCEPTION
+                log.info(
+                    "PartitionProcessor of EventProcessor instance %r of eventhub %r partition %r consumer group %r "
+                    "has met an exception receiving events. It's being closed. The exception is %r.",
+                    partition_processor._checkpoint_manager._instance_id,
+                    partition_processor._eventhub_name,
+                    partition_processor._partition_id,
+                    partition_processor._consumer_group_name,
+                    eh_err
+                )
+                await partition_processor.close(reason=reason)
+                break
+            try:
+                await partition_processor.process_events(events)
+            except Exception as exp:  # user code has caused an error
+                log.info(
+                    "PartitionProcessor of EventProcessor instance %r of eventhub %r partition %r consumer group %r "
+                    "has met an exception from user code process_events. It's being closed. The exception is %r.",
+                    partition_processor._checkpoint_manager._instance_id,
+                    partition_processor._eventhub_name,
+                    partition_processor._partition_id,
+                    partition_processor._consumer_group_name,
+                    exp
+                )
+                await partition_processor.close(reason=CloseReason.USER_EXCEPTION)
+                break
+        else:
```

Comment:
We can remove **else**. (Is it an **else** statement matching the while loop?)
After the while loop ends, we just continue to do the close process.

Reply:
The while loop has "break". When it breaks, the else is skipped. When it doesn't break, the else is not skipped.

Reply:
Preview 3 code no longer has it